firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +313,258 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
+
+        This is temporary while we wait for structure-preserving tfsc kernels.
+        """
+        self.prolong_kernel = self.prolongation_transfer_kernel_action(Vf, self.uc)
         matrix_kernel = self.prolongation_transfer_kernel_action(Vf, firedrake.TestFunction(Vc))
+        # The way we transpose the prolongation kernel is suboptimal.
+        # A local matrix is generated each time the kernel is executed.
         element_kernel = loopy.generate_code_v2(matrix_kernel.code).device_code()
         element_kernel = element_kernel.replace("void expression_kernel", "static void expression_kernel")
         dimc = Vc.finat_element.space_dimension() * Vc.value_size
         dimf = Vf.finat_element.space_dimension() * Vf.value_size
-        self.restrict_code = f"""
-{element_kernel}
+        restrict_code = f"""
+        {element_kernel}
+
+        void restriction(double *restrict Rc, const double *restrict Rf, const double *restrict w)
+        {{
+            double Afc[{dimf}*{dimc}] = {{0}};
+            expression_kernel(Afc);
+            for (int32_t i = 0; i < {dimf}; i++)
+               for (int32_t j = 0; j < {dimc}; j++)
+                   Rc[j] += Afc[i*{dimc} + j] * Rf[i] * w[i];
+        }}
+        """
+        self.restrict_kernel = op2.Kernel(restrict_code, "restriction")
 
+    @staticmethod
+    def prolongation_transfer_kernel_action(Vf, expr):
+        from tsfc import compile_expression_dual_evaluation
+        from tsfc.fiatinterface import create_element
+        coords = Vf.ufl_domain().coordinates
+        to_element = create_element(Vf.ufl_element(), vector_is_mixed=False)
+        ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, to_element, coords, coffee=False)
+        return op2.Kernel(ast, ast.name)
 
-void restriction(double *restrict Rc, const double *restrict Rf, const double *restrict w)
-{{
-    double Afc[{dimf}*{dimc}] = {{0}};
-    expression_kernel(Afc);
-    for (int32_t i = 0; i < {dimf}; i++)
-       for (int32_t j = 0; j < {dimc}; j++)
-           Rc[j] += Afc[i*{dimc} + j] * Rf[i] / w[i];
-}}
-"""
+    def make_blas_kernels(self, Vf, Vc):
+        """
+        Interpolation and restriction kernels between CG/DG
+        tensor product spaces on quads and hexes.
 
-        self.restrict_kernel = op2.Kernel(self.restrict_code, "restriction")
-        self.mesh = Vf.mesh()
+        Works by tabulating the coarse 1D Lagrange basis
+        functions in the (Nf+1)-by-(Nc+1) matrix Jhat,
+        and using the fact that the 2d/3d tabulation is the
+        tensor product J = kron(Jhat, kron(Jhat, Jhat))
+        """
+        ndim = Vf.ufl_domain().topological_dimension()
+        zf = self.get_nodes_1d(Vf)
+        zc = self.get_nodes_1d(Vc)
+        Jnp = self.barycentric(zc, zf)
+        # Declare array shapes to be used as literals inside the kernels, I follow to the m-by-n convention.
+        (mx, nx) = Jnp.shape
+        (my, ny) = (mx, nx)
```

Comment:
```suggestion
        (my, ny) = (mx, nx) if ndim >= 2 else (1, 1)
```

This is needed otherwise you walk off the end of some arrays for 1D prolongation/restriction.

Reply:
For 1D, the standard kernels are equivalent to these tensor product kernels. So should we return `False` in tensor_product_space_query for that case? I think the above change should be done anyway though.